Make StoreBarrierInsertion aware of the implications of concurrent marking

saambarati opened the pull request:
#### 4724c2bd9edbf74eaf805633cd98859cfa8f974c
<pre>
Make StoreBarrierInsertion aware of the implications of concurrent marking
<a href="https://bugs.webkit.org/show_bug.cgi?id=242046">https://bugs.webkit.org/show_bug.cgi?id=242046</a>

Reviewed by Yusuke Suzuki.

We need to consider nodes that might leak objects we&apos;ve allocated into the heap.
Once an object is leaked, we can no longer elide barriers on it.
Let&apos;s motivate this requirement with an example:
D@30: JSConstant(Int32: 42)
D@35: GetStack(arg1)
D@21: CheckStructure(Cell:D@35, [%ED:Object])
D@23: GetStack(arg2)
D@25: NewObject()
D@33: PutByOffset(KnownCell:D@25, KnownCell:D@25, Check:Untyped:Kill:D@30, id0{x})
D@34: PutStructure(KnownCell:D@25, %DN:Object -&gt; %Ch:Object)
D@40: PutByOffset(KnownCell:D@35, KnownCell:D@35, Check:Untyped:D@25, id1{p})
D@45: FencedStoreBarrier(Check:KnownCell:Kill:D@35)
&lt;-- P1
D@41: PutByOffset(KnownCell:D@25, KnownCell:D@25, Check:Untyped:Kill:D@23, id2{y})
&lt;-- P2

Let&apos;s say at the program point P1, the barrier @45 didn&apos;t fire because @35 is already grey.
Because @35 is grey, at P1, let&apos;s say the concurrent marker marks and traces @35, and also
marks and traces @25. So at P1, the concurrent marker blackens @35 and @25.
Now, let&apos;s consider program point P2.
If we didn&apos;t barrier @25 at P2, we will never see that @25 points to @23, because @25 is already
black. This is because after @25 was allocated, it escaped into the heap (at @40). Once an allocation
escapes into the heap, it can be blackened at any point by the concurrent marker.
So this analysis must mark an allocation that escapes to the heap as being part of the primordial
epoch.

StoreBarrierInsertion didn&apos;t used to insert a barrier at P2. This patch
teaches it that it must do so.

* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:

Canonical link: <a href="https://commits.webkit.org/251925@main">https://commits.webkit.org/251925@main</a>
</pre>

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
```diff
@@ -402,6 +403,65 @@ class StoreBarrierInsertionPhase : public Phase {
+                    case PutClosureVar:
+                        escape(m_node->child2().node());
+                        break;
+                    default:
```
I wonder if we should add more nodes here (at least, nodes listed in the former switch) to avoid clearing epoch for base.

saambarati replied:
I'm not sure there are really any other nodes that would make a difference here. They'd need to not GC, and write into something other than side state, and be in a place where they take allocation nodes as an argument to the node.

saambarati committed:
Make StoreBarrierInsertion aware of the implications of concurrent marking
https://bugs.webkit.org/show_bug.cgi?id=242046

Reviewed by Yusuke Suzuki.

We need to consider nodes that might leak objects we've allocated into the heap.
Once an object is leaked, we can no longer elide barriers on it.
Let's motivate this requirement with an example:
D@30: JSConstant(Int32: 42)
D@35: GetStack(arg1)
D@21: CheckStructure(Cell:D@35, [%ED:Object])
D@23: GetStack(arg2)
D@25: NewObject()
D@33: PutByOffset(KnownCell:D@25, KnownCell:D@25, Check:Untyped:Kill:D@30, id0{x})
D@34: PutStructure(KnownCell:D@25, %DN:Object -> %Ch:Object)
D@40: PutByOffset(KnownCell:D@35, KnownCell:D@35, Check:Untyped:D@25, id1{p})
D@45: FencedStoreBarrier(Check:KnownCell:Kill:D@35)
<-- P1
D@41: PutByOffset(KnownCell:D@25, KnownCell:D@25, Check:Untyped:Kill:D@23, id2{y})
<-- P2

Let's say at the program point P1, the barrier @45 didn't fire because @35 is already grey.
Because @35 is grey, at P1, let's say the concurrent marker marks and traces @35, and also
marks and traces @25. So at P1, the concurrent marker blackens @35 and @25.
Now, let's consider program point P2.
If we didn't barrier @25 at P2, we will never see that @25 points to @23, because @25 is already
black. This is because after @25 was allocated, it escaped into the heap (at @40). Once an allocation
escapes into the heap, it can be blackened at any point by the concurrent marker.
So this analysis must mark an allocation that escapes to the heap as being part of the primordial
epoch.

StoreBarrierInsertion didn't used to insert a barrier at P2. This patch
teaches it that it must do so.

* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:

Canonical link: https://commits.webkit.org/251925@main
